File: source/javascript/binary_search_iterative.js
// Binary Search (Iterative)
function binarySearch(arr, target) {
    let low = 0, high = arr.length - 1;
    while (low <= high) {
        const mid = low + Math.floor((high - low) / 2);
        if (arr[mid] === target) return mid;
        else if (arr[mid] < target) low = mid + 1;
        else high = mid - 1;
    }
    return null;
}

const arr = [2, 5, 8, 12, 16, 23, 38, 56, 72, 91];

const result1 = binarySearch(arr, 23);
if (result1 !== null) console.log(`search 23: found at index ${result1}`);
else console.log("search 23: not found");

const result2 = binarySearch(arr, 50);
if (result2 !== null) console.log(`search 50: found at index ${result2}`);
else console.log("search 50: not found");

File: source/rust/binary_search_iterative.rs
fn binary_search(arr: &[i32], target: i32) -> Option<usize> {
    let mut low = 0i64;
    let mut high = arr.len() as i64 - 1;
    while low <= high {
        let mid = low + (high - low) / 2;
        if arr[mid as usize] == target {
            return Some(mid as usize);
        } else if arr[mid as usize] < target {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    None
}

fn main() {
    let arr = [2, 5, 8, 12, 16, 23, 38, 56, 72, 91];
    match binary_search(&arr, 23) {
        Some(i) => println!("search 23: found at index {}", i),
        None => println!("search 23: not found"),
    }
    match binary_search(&arr, 50) {
        Some(i) => println!("search 50: found at index {}", i),
        None => println!("search 50: not found"),
    }
}
